### PR TITLE
feat: Register user

### LINKED ISSUE #38: Create UI for participant enrollment
**Is your feature request related to a problem? Please describe.**

We need a way to identify users to be able to associate the UUID assigned to the music playing data with a user identify - for example, an email address. This is required to cross-reference the music playing data with other data on the caregiver/receiver collected via other methods (e.g., web survey).

For now, we can target single caregiver / care receiver model.

Then, the app should log usage data using this UUID to Firebase (see https://github.com/CUTR-at-USF/MUSER/issues/17 for details on this). The UUID to email mapping will be stored in a Google Sheet (see https://github.com/CUTR-at-USF/MUSER/issues/39 for details on this).

This issue will focus on implementing the UI for user enrollment.

**Describe the solution you'd like**
We've implemented something similar in the OneBusAway open-source app:
https://github.com/OneBusAway/onebusaway-android/

Example prompt for user registration in the OneBusAway:
https://github.com/OneBusAway/onebusaway-android/blob/master/onebusaway-android/src/main/java/org/onebusaway/android/ui/HomeActivity.java#L396

We store email->UUID mapping in Google Sheet, and we store travel behavior info w/ UUID in Firebase. More info in presentation and final report from project:

Report - https://scholarcommons.usf.edu/cgi/viewcontent.cgi?article=1012&context=cutr_nctr
Slides - https://www.slideshare.net/sjbarbeau/improving-the-quality-and-cost-effectiveness-of-multimodal-travel-behavior-data-collection

### PR DESCRIPTION
**Pull request by BumbleFlash**
This PR is to create dialogs to enroll users who would like to participate in the study. 

Changes include:-

- [x] Create a participation dialog to ask if the user would like to enroll themselves in the study. 
- [x] Show the Informed Consent to the user if they agree to enroll in the study. 
- [x] Create a dialog to prompt the user to enter their email address.
- [x] Call the `registerUser` function to enroll them into the study. 
- [x] Integrate the UserRegistrationManager.java to the `BaseActivity` of the app. 
- [x] Save UUID-> Email mapping to Google App Scripts
- [x] Remove the alert from `HomeActivity` and place it after permissions pop up. 
- [x] Remove `ic_light_bulb`
- [x] Replace `AlertDialog` with `MaterialDialog` 
- [x] Add participation required dialog when the user doesn't want to enroll in the study and exit the app. 

"Closes #38, #39"

### CODE REVIEW

**Comment by barbeau** [comment]
@BumbleFlash I merged all the Firebase code into `usf` via https://github.com/CUTR-at-USF/MUSER/pull/66, so you may need to update this branch to reflect those changes (hopefully few conflicts).

**Comment by BumbleFlash** [comment]
@barbeau I have merged the `usf `and I think this should have resolved any conflicts for now.

**Comment by barbeau** [comment]
@BumbleFlash Ok, great - should this be converted from draft then to "ready for review"? Or are you still working on it?

File: app/src/main/res/drawable-anydpi/ic_light_bulb.xml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<vector xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (barbeau) [inline app/src/main/res/drawable-anydpi/ic_light_bulb.xml]: Let's remove the light-bulb icon (all of them) and just use the app icon instead.

File: app/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -1,24 +1,19 @@
 <?xml version="1.0" encoding="utf-8"?>
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
     package="com.simplecity.amp_library">
-
     <!-- ## 'Protection-Normal' permissions ## -->
```
Comment (barbeau) [inline app/src/main/AndroidManifest.xml]: Looks like a lot of unrelated (formatting?) changes here - could you please remove them?

File: app/src/main/java/edu/usf/sas/pal/muser/ui/HomeActivity.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+import edu.usf.sas.pal.muser.manager.UserRegistrationManager;
+
+public class HomeActivity extends AppCompatActivity {
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_home);
+        new UserRegistrationManager(getApplicationContext(), this)
```
Comment (barbeau) [inline app/src/main/java/edu/usf/sas/pal/muser/ui/HomeActivity.java]: Let's try to re-work this to remove HomeActivity. So the file permission pops up first, and the `new UserRegistrationManager(getApplicationContext(), this).registerParticipant()` is called from the permissions granted callback.

Currently there are issues with the Activity backstack and the blank HomeActivity hanging around when it shouldn't. This will be a constant hassle to deal with and create some weird UI experiences on some devices.

File: app/src/main/java/edu/usf/sas/pal/muser/manager/UserRegistrationManager.java [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+    }
+
+    public void registerParticipant(boolean forceStart){
+        // TODO check possible limit of enrollment from a certain location?
+        boolean isUserOptOut = PreferenceUtils.getBoolean(EventConstants.USER_OPT_OUT,
+                false);
+        if (forceStart) isUserOptOut = false;
+
+        if(isUserOptOut){
+         switchToMainActivity(mActivityContext);
```
Comment (barbeau) [inline app/src/main/java/edu/usf/sas/pal/muser/manager/UserRegistrationManager.java]: These need to be indented more, and also missing a space between `if` and `(`

File: app/src/main/java/edu/usf/sas/pal/muser/manager/UserRegistrationManager.java [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+
+    public void registerParticipant(){
+        registerParticipant(false);
+    }
+
+    public void registerParticipant(boolean forceStart){
+        // TODO check possible limit of enrollment from a certain location?
+        boolean isUserOptOut = PreferenceUtils.getBoolean(EventConstants.USER_OPT_OUT,
+                false);
+        if (forceStart) isUserOptOut = false;
```
Comment (barbeau) [inline app/src/main/java/edu/usf/sas/pal/muser/manager/UserRegistrationManager.java]: Could you please wrap all IF statement bodies, even single lines, with `{` and `}`. I know this is a style thing but I prefer it this way.

File: app/src/main/java/edu/usf/sas/pal/muser/manager/UserRegistrationManager.java [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+        mActivityContext = activityContext;
+        mApplicationContext = activityContext;
+    }
+
+    public void registerParticipant(){
+        registerParticipant(false);
+    }
+
+    public void registerParticipant(boolean forceStart){
+        // TODO check possible limit of enrollment from a certain location?
```
Comment (barbeau) [inline app/src/main/java/edu/usf/sas/pal/muser/manager/UserRegistrationManager.java]: What does this TODO refer to?

File: app/src/main/java/edu/usf/sas/pal/muser/manager/UserRegistrationManager.java [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+    public UserRegistrationManager(Context applicationContext, Context activityContext){
+        mActivityContext = activityContext;
+        mApplicationContext = activityContext;
+    }
+
+    public void registerParticipant(){
+        registerParticipant(false);
+    }
+
+    public void registerParticipant(boolean forceStart){
```
Comment (barbeau) [inline app/src/main/java/edu/usf/sas/pal/muser/manager/UserRegistrationManager.java:53]: Can you document what `forceStart` does here in Javadocs?

File: app/src/main/java/edu/usf/sas/pal/muser/manager/UserRegistrationManager.java [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+        }
+
+        boolean isUserOptIn = PreferenceUtils.getBoolean(EventConstants.USER_OPT_IN,
+                false);
+        if ((!isUserOptIn)) {
+            showParticipationDialog();
+        } else {
+            switchToMainActivity(mActivityContext);
+        }
+
```
Comment (barbeau) [inline app/src/main/java/edu/usf/sas/pal/muser/manager/UserRegistrationManager.java]: Extra space

File: app/src/main/java/edu/usf/sas/pal/muser/util/FirebaseIOUtils.java [hunk 9]
```diff
@@ -72,21 +80,55 @@ private static synchronized String getRecordID(){
 
     public static void logErrorMessage(Exception e, String message) {
         if (e != null) {
-            Log.d(TAG, message + e.getMessage());
+            Log.d(TAG, message +
```
Comment (barbeau) [inline app/src/main/java/edu/usf/sas/pal/muser/util/FirebaseIOUtils.java]: This can all be on one line

File: app/src/main/java/edu/usf/sas/pal/muser/util/PreferenceUtils.java [hunk 10]
```diff
@@ -39,4 +39,19 @@ public static void saveLong(String key, long value) {
     public static long getLong(String key, long defaultValue) {
         return ShuttleApplication.getPrefs().getLong(key, defaultValue);
     }
+
```
Comment (barbeau) [inline app/src/main/java/edu/usf/sas/pal/muser/util/PreferenceUtils.java:40]: Extra lines

File: app/src/main/res/layout/research_participation_dialog.xml [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+        android:text="@string/register_user_opt_in_message"
+        android:textSize="16sp"
+        android:autoLink="all"
+        android:linksClickable="true" />
+
+    <CheckBox
+        android:id="@+id/research_never_ask_again"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:text="@string/never_ask_again_message" />
```
Comment (barbeau) [inline app/src/main/res/layout/research_participation_dialog.xml:42]: You can add `android:visibility="gone"` to this to hide the view, and then you can remove the programmatic reference to it as well. We aren't going to let people opt-out for now.

File: app/src/main/res/values/strings.xml [hunk 12]
```diff
@@ -285,10 +285,10 @@
     <string name="pref_string_title_themes">Themes</string>
     <!-- Setting title for set default theme preference-->
     <!-- TODO to translate -->
-    <string name = "pref_string_set_default_theme">Set Default theme</string>
+    <string name="pref_string_set_default_theme">Set Default theme</string>
     <!-- Summary of set default theme preference-->
     <!-- TODO to translate -->
-    <string name = "pref_summary_set_default_theme">Sets the theme of the app back to the default MUSER color</string>
```
Comment (barbeau) [inline app/src/main/res/values/strings.xml:291]: Unrelated changes

File: app/src/main/res/values/strings.xml [hunk 13]
```diff
@@ -841,7 +841,22 @@
     <string name="saf_show_files_button">Show files</string>
 
 
-
     <string name="notification_text_shuttle_running">Shuttle service running…</string>
 
+
```
Comment (barbeau) [inline app/src/main/res/values/strings.xml]: Extra line

**Review by barbeau — CHANGES_REQUESTED** [review]
@BumbleFlash Thanks for working on this! Generally I think most of the code is good, but we need to change it to remove the `HomeActivity` for reasons mentioned below.

**Comment by barbeau** [comment]
One more item - The "No" and "Yes" buttons are very lightly colored - is that the theme accent color?

image

File: app/src/main/java/edu/usf/sas/pal/muser/manager/UserRegistrationManager.java (hunk 6 above)
Comment (BumbleFlash) [reply app/src/main/java/edu/usf/sas/pal/muser/manager/UserRegistrationManager.java]: While I was reading through the OBA code, I saw a section of code that was keeping track of enrollments from a certain location. So I added this TODO to flag this portion if we were to add something like that in future.

File: app/src/main/java/edu/usf/sas/pal/muser/manager/UserRegistrationManager.java (hunk 7 above)
Comment (BumbleFlash) [reply app/src/main/java/edu/usf/sas/pal/muser/manager/UserRegistrationManager.java:53]: Sounds good. I haven't added Javadocs to any of these functions yet. I'll have them ready as I complete the checklist?

File: app/src/main/java/edu/usf/sas/pal/muser/util/FirebaseIOUtils.java (hunk 9 above)
Comment (BumbleFlash) [reply app/src/main/java/edu/usf/sas/pal/muser/util/FirebaseIOUtils.java]: I guess most of these changes are from the `register-user` branch cause I remember fixing them in the `firebase` branch

**Comment by BumbleFlash** [comment]
I'm guessing the `AlertDialog` uses the default color as the `primaryColor` for the texts.  This is one way to change it forcefully. 
https://stackoverflow.com/questions/33437398/how-to-change-textcolor-in-alertdialog

**Comment by barbeau** [comment]
Ah - I think the issue for the AlertDialog is that the import is for `android.app.AlertDialog` instead of `androidx.appcompat.app.AlertDialog`. But we can't use AndroidX due to https://github.com/CUTR-at-USF/MUSER/issues/25. So could you try using `android.support.v7.app.AlertDialog` instead?

File: app/src/main/java/edu/usf/sas/pal/muser/manager/UserRegistrationManager.java (hunk 6 above)
Comment (barbeau) [reply app/src/main/java/edu/usf/sas/pal/muser/manager/UserRegistrationManager.java]: Oh, gotcha - you can remove this, as we don't have an equivalent regional server architecture here with MUSER

**Comment by BumbleFlash** [comment]
Sounds good. Sorry I requested your review before seeing your latest comments. I'll try to implement them as you review the code.

**Comment by BumbleFlash** [comment]
@barbeau I tried importing the v7 of the `AlertDialog`, the text color looks unchanged to me. Do you want me to implement it as per the above StackOverflow link?

**Comment by barbeau** [comment]
Setting the color using HTML fonts seems really hacky to me. Could you try adding the material design library and using this instead:
https://material.io/develop/android/components/dialogs

**Comment by barbeau** [comment]
I think the issue is that this is picking up the `colorAccent` color from the theme, so you may have the same issue with the material design library. The accent color is `#B5D7DE`, which is really light. We might need to pick a darker accent color. If you get the same result with the material design library, let's just leave it as-is in the current PR now (without material design library) and we'll take it back to the team for a color change, because it will be an issue elsewhere too.

**Comment by BumbleFlash** [comment]
@barbeau Before importing the `MaterialDaiog`, I noticed that `Aesthetic `was already overriding and using the library. Although by default, the dialog was picking up the `AccentColor`, I was able to change it to a darker color programmatically by using the `.positiveColor` function while building the dialog. Here's how it looks. Should we stick to this or would you like me to import the `MaterialDialog`? 



Here's the code:-
     
    new MaterialDialog.Builder(mActivityContext)
                .customView(view, false)
                .title(R.string.register_user_opt_in_title)
                .icon(createIcon())
                .limitIconToDefaultSize()
                .cancelable(false)
                .positiveText(R.string.research_user_dialog_yes)
                .positiveColor(mActivityContext.getResources().getColor(R.color.colorPrimaryDark))
                .onPositive(((dialog, which) -> showInformedConsent()))
                .negativeText(R.string.research_user_dialog_no)
                .onNegative(((dialog, which) -> {
                    if (neverShowDialog.isChecked()){
                        optOutUser();
                    }
                }))
                .negativeColor(mActivityContext.getResources().getColor(R.color.colorPrimaryDark))
                .build()
                .show();

Note: The checkbox will be removed as I commit this change if approved.

**Comment by barbeau** [comment]
@BumbleFlash That looks great, let's go with that!

File: app/src/main/java/edu/usf/sas/pal/muser/manager/UserRegistrationManager.java [hunk 14]
```diff
@@ -0,0 +1,222 @@
...
+                .icon(createIcon())
+                .limitIconToDefaultSize()
+                .cancelable(false)
+                .positiveText(R.string.research_user_dialog_yes)
+                .positiveColor(mActivityContext.getResources().getColor(R.color.colorPrimaryDark))
+                .onPositive(((dialog, which) -> showInformedConsent()))
+                .negativeText(R.string.research_user_dialog_no)
+                .onNegative(((dialog, which) -> {
+                    if (neverShowDialog.isChecked()) {
+                        optOutUser();
```
Comment (barbeau) [inline app/src/main/java/edu/usf/sas/pal/muser/manager/UserRegistrationManager.java]: For now, this should show a Toast and exit the app saying "Sorry you need to enroll in the research study to use the app" or something similar. I believe we shouldn't opt out the user, because we want them to be prompted again if they re-start the app, so their only option is to enroll if they want to use the app (i.e., we keep prompting them until they enroll).

File: app/src/main/java/com/simplecity/amp_library/ShuttleApplication.java [hunk 15]
```diff
@@ -107,11 +107,9 @@ public void onCreate() {
             return;
         }
 
-        String email = "a@a.com";
         // TODO
```
Comment (barbeau) [inline app/src/main/java/com/simplecity/amp_library/ShuttleApplication.java:110]: TODO should be removed if this doesn't happen here, and if the feature isn't implemented yet we should move the TODO to where it should be implemented.

File: app/src/main/AndroidManifest.xml [hunk 16]
```diff
@@ -33,40 +33,36 @@
 
         <meta-data
```
Comment (barbeau) [inline app/src/main/AndroidManifest.xml]: I think all of these changes to the manifest can be reverted? They look like automated reformatting.

File: app/build.gradle [hunk 17]
```diff
@@ -236,6 +236,8 @@ dependencies {
     implementation Dependencies.Square.retrofit
     implementation Dependencies.Square.retrofitGson
     implementation Dependencies.Square.sqlBrite
+    implementation 'com.android.support:appcompat-v7:28.0.0'
```
Comment (barbeau) [inline app/build.gradle]: Can these new dependencies be removed?

File: app/src/main/java/com/simplecity/amp_library/ui/screens/widgets/WidgetFragment.java [hunk 18]
```diff
@@ -39,7 +39,7 @@ public void onAttachFragment(Fragment childFragment) {
     /**
      * Creates a new instance of the {@link WidgetFragment}
      *
-     * @param widgetLayoutResId the id of the layout to use for the widget
+     * @param widgetLayoutResId the id of the research_participation_dialog to use for the widget
```
Comment (barbeau) [inline app/src/main/java/com/simplecity/amp_library/ui/screens/widgets/WidgetFragment.java]: Was this an intentional change? If so, why?

File: app/src/main/java/edu/usf/sas/pal/muser/manager/UserRegistrationManager.java [hunk 19]
```diff
@@ -0,0 +1,222 @@
...
+                .cancelable(false)
+                .positiveText(R.string.participation_consent_agree)
+                .positiveColor(mActivityContext.getResources().getColor(R.color.colorPrimaryDark))
+                .onPositive((dialog, which) -> {
+                    showEmailDialog();
+                })
+                .negativeText(R.string.participation_consent_disagree)
+                .negativeColor(mActivityContext.getResources().getColor(R.color.colorPrimaryDark))
+                .onNegative((dialog, which) -> {
+                    optOutUser();
```
Comment (barbeau) [inline app/src/main/java/edu/usf/sas/pal/muser/manager/UserRegistrationManager.java]: Similar here - for now, this should show a Toast and exit the app saying "Sorry you need to agree to the informed consent to use the app" or something similar. I believe we shouldn't opt out the user, because we want them to be prompted again if they re-start the app, so their only option is to enroll if they want to use the app (i.e., we keep prompting them until they enroll).

File: app/src/main/res/values/strings.xml [hunk 20]
```diff
@@ -841,7 +841,21 @@
     <string name="saf_show_files_button">Show files</string>
 
 
-
     <string name="notification_text_shuttle_running">Shuttle service running…</string>
 
+    <!--User Registration-->
+    <string name="register_user_opt_in_message">We\'d like your help with research to improve the music service.</string>
+    <string name="never_ask_again_message"> Never ask again</string>
+    <string name="register_user_opt_in_title"> Help us improve MUSER!</string>
```
Comment (barbeau) [inline app/src/main/res/values/strings.xml]:
```suggestion
    <string name="register_user_opt_in_title">Participate in research</string>
```

File: app/src/main/res/values/strings.xml [hunk 21]
```diff
@@ -841,7 +841,21 @@
     <string name="saf_show_files_button">Show files</string>
 
 
-
     <string name="notification_text_shuttle_running">Shuttle service running…</string>
 
+    <!--User Registration-->
+    <string name="register_user_opt_in_message">We\'d like your help with research to improve the music service.</string>
```
Comment (barbeau) [inline app/src/main/res/values/strings.xml]:
```suggestion
    <string name="register_user_opt_in_message">Currently MUSER is only available for participants in USF research. Would you like to join the study?</string>
```

File: app/src/main/res/values/strings.xml [hunk 22]
```diff
@@ -841,7 +841,21 @@
...
+    <string name="never_ask_again_message"> Never ask again</string>
+    <string name="register_user_opt_in_title"> Help us improve MUSER!</string>
+    <string name="research_user_dialog_yes">Yes</string>
+    <string name="research_user_dialog_no">No</string>
+    <string name="participation_consent_agree">Agree</string>
+    <string name="participation_consent_disagree">Disagree</string>
+    <string name="research_email_hint">Email</string>
+    <string name="research_confirm_email_hint">Confirm Email</string>
+    <string name="research_email_message">Please provide your email address for the online surveys</string>
```
Comment (barbeau) [inline app/src/main/res/values/strings.xml]:
```suggestion
    <string name="research_email_message">Please provide your email address for online surveys and communication with the research team</string>
```

File: app/src/main/res/values/donottranslate.xml [hunk 23]
```diff
@@ -32,4 +32,8 @@
         <item>@string/playlists_title</item>
     </string-array>
 
+    <!-- Do not translate. URL for mapping users to google App Scripts.-->
```
Comment (barbeau) [inline app/src/main/res/values/donottranslate.xml]:
```suggestion
    <!-- Do not translate. URLs for Google App Scripts that have ID/email mapping-->
```

File: app/src/main/java/edu/usf/sas/pal/muser/manager/UserRegistrationManager.java [hunk 24]
```diff
@@ -0,0 +1,222 @@
...
+import java.net.HttpURLConnection;
+import java.net.URL;
+
+import edu.usf.sas.pal.muser.constants.EventConstants;
+import edu.usf.sas.pal.muser.util.FirebaseIOUtils;
+import edu.usf.sas.pal.muser.util.PreferenceUtils;
+
+/**
+ * Manager class to handle user registration. Creates dialogs to prompt the user about the study
+ * fetch  email if user agrees to enroll.
```
Comment (barbeau) [inline app/src/main/java/edu/usf/sas/pal/muser/manager/UserRegistrationManager.java]:
```suggestion
 * Manager class to handle user registration. Creates dialogs to prompt the user about the study
 * and collect their email if user agrees to enroll.
```

File: app/src/main/java/edu/usf/sas/pal/muser/util/FirebaseIOUtils.java [hunk 25]
```diff
@@ -73,20 +81,55 @@ private static synchronized String getRecordID(){
...
+                        new AsyncTask<Void, Integer, Integer>(){
+
+                            @Override
+                            protected Integer doInBackground(Void... voids) {
+                                int responseCode = 0;
+                                try {
+                                     responseCode = UserRegistrationManager
+                                            .saveEmailAddress(firebaseAuth.getUid(),
+                                            email);
+
```
Comment (barbeau) [inline app/src/main/java/edu/usf/sas/pal/muser/util/FirebaseIOUtils.java]: Extra line

File: app/src/main/java/edu/usf/sas/pal/muser/util/FirebaseIOUtils.java [hunk 26]
```diff
@@ -73,20 +81,55 @@ private static synchronized String getRecordID(){
...
+    public static void registerUser(String email, Context context){
         FirebaseAuth firebaseAuth = FirebaseAuth.getInstance();
         firebaseAuth.signInAnonymously()
                 .addOnCompleteListener(task -> {
                     if (task.isSuccessful()) {
                         Log.d(TAG, "Firebase user initialized with id:" + firebaseAuth.getUid());
-                        // TODO save email address to Google APP Scripts
-                        PreferenceUtils.saveString(EventConstants.USER_ID, firebaseAuth.getUid());
+                        new AsyncTask<Void, Integer, Integer>(){
+
```
Comment (barbeau) [inline app/src/main/java/edu/usf/sas/pal/muser/util/FirebaseIOUtils.java]: Extra line

File: app/src/main/res/values/strings.xml [hunk 27]
```diff
@@ -841,7 +841,21 @@
     <string name="saf_show_files_button">Show files</string>
 
 
-
     <string name="notification_text_shuttle_running">Shuttle service running…</string>
 
+    <!--User Registration-->
+    <string name="register_user_opt_in_message">We\'d like your help with research to improve the music service.</string>
+    <string name="never_ask_again_message"> Never ask again</string>
```
Comment (barbeau) [inline app/src/main/res/values/strings.xml]: There is an extra space at the start of this and the following string.

File: app/src/main/java/edu/usf/sas/pal/muser/manager/UserRegistrationManager.java [hunk 28]
```diff
@@ -0,0 +1,222 @@
...
+    private void showEmailDialog(String email){
+        LayoutInflater inflater = ((Activity) mActivityContext).getLayoutInflater();
+        final View editTextView = inflater.inflate(R.layout.research_email_dialog, null);
+        EditText emailEditText = editTextView.findViewById(R.id.tb_email_edittext);
+        EditText emailEditTextConfirm = editTextView.findViewById(R.id.tb_email_edittext_confirm);
+
+        if (email != null) {
+            emailEditText.setText(email);
+        }
+        new MaterialDialog.Builder(mActivityContext)
```
Comment (barbeau) [inline app/src/main/java/edu/usf/sas/pal/muser/manager/UserRegistrationManager.java:153]: This dialog shouldn't be cancelable - needs `.setCancelable(false)`

**Review by barbeau — CHANGES_REQUESTED** [review]
@BumbleFlash Thanks! This mostly looks good - a few suggestions in-line, and comments for how the "No" and "Decline" buttons should work now (considering that we're forcing users to enroll in the study to use the app).

File: app/src/main/java/edu/usf/sas/pal/muser/manager/UserRegistrationManager.java (hunk 14 above)
Comment (BumbleFlash) [reply app/src/main/java/edu/usf/sas/pal/muser/manager/UserRegistrationManager.java]: Thank you for your feedback. I understand. One question about exiting the app, won't that be a bit inconvenient for the user if they don't want to enroll in the app? From a naive user's POV, I open the app and I say no to the prompt, the app kinda quits in a rude manner. Alternatively, I can keep prompting the dialog box if the user keeps saying "No". Let me know what you think.

File: app/src/main/AndroidManifest.xml (hunk 16 above)
Comment (BumbleFlash) [reply app/src/main/AndroidManifest.xml]: The manifest is pulled from the `dev` branch. Since the `HomeActivity` doesn't exist anymore, I took the chance to just pull the one from Shuttle.

File: app/src/main/java/com/simplecity/amp_library/ui/screens/widgets/WidgetFragment.java (hunk 18 above)
Comment (BumbleFlash) [reply app/src/main/java/com/simplecity/amp_library/ui/screens/widgets/WidgetFragment.java]: Long story short, I accidentally spammed the "next" button while creating the dialog, and the default filename was set to`layout.xml`. When I refactored the file to rename to the `research_participation_dialog`, AS acted up and changed whatever that said `layout` to `research_participation_dialog`. I thought I rolled back everything that didn't make sense but I guess there is still some left to change.

File: app/src/main/java/edu/usf/sas/pal/muser/manager/UserRegistrationManager.java (hunk 14 above)
Comment (barbeau) [reply app/src/main/java/edu/usf/sas/pal/muser/manager/UserRegistrationManager.java]: That's a reasonable alternative as long as there is clear messaging in the app, although the concern with an infinite loop is that there is that there is no obvious way for the user to exit the app from within it.

How about this - if the user answers no/decline, show another uncancelable dialog with a single "OK" positive button with the following:
* Title - "Participation required"
* Message - "Currently all users of this app are required to enroll in the research study. The app will now exit. If you'd like to enroll, please restart the app."

And when you tap on "OK" the app exits.

This is a little more graceful than a toast but provides more direct messaging to the user.

Comment (BumbleFlash) [reply app/src/main/java/edu/usf/sas/pal/muser/manager/UserRegistrationManager.java]: This sounds perfect. I'll have it ready ASAP.

File: app/build.gradle (hunk 17 above)
Comment (BumbleFlash) [reply app/build.gradle]: Honestly, I don't remember adding them to the Gradle file. Did AS just add the dependency when I imported `v7.AlertDialog`?

Comment (barbeau) [reply app/build.gradle]: Probably - AS is becoming "smarter" about adding dependencies when needed.

**Comment by BumbleFlash** [comment]
@barbeau Does this look good? Also, do you want the button on the right or left? Let me know if you'd like me to make any changes.

**Comment by barbeau** [comment]
Looks good! Whatever the normal "positive" button position is, which I believe is the left?

**Review by barbeau — APPROVED** [review]
Looks great @BumbleFlash, thanks for all your work on this! I added a few commits for additional cleanup that was faster than going through another review cycle, but feel free to take a look and let me know if you have questions about any of it.

**Comment by BumbleFlash** [comment]
@barbeau LGTM - Thanks for taking the time to review and to clean up the code 👍
